app/src/main/java/com/github/ybecker/epforuml/database/FirebaseDatabaseAdapter.kt

```diff
@@ -1,59 +1,345 @@
+                    }
+                }
+                future.complete(questions)
+            }.addOnFailureListener {
+            future.completeExceptionally(it)
+        }
+
+        return future.get(5, TimeUnit.SECONDS)
     }
 
-    override fun getAnswersForQuestion(question: Question): Set<Answer> {
-        TODO("Not yet implemented")
+    //Note that using question.answers in the main is false because you don't take new values in the db into account !
+    override fun getQuestionAnswers(question: Question): Set<Answer> {
+        val future = CompletableFuture<Set<Answer>>()
+        // go in "question/questionId" dir
+        db.child(questionsPath).child(question.questionId).child(answersPath).get().addOnSuccessListener {
+            val answers = mutableSetOf<Answer>()
+            // add every question's answer that is not null in the map
+            for (courseSnapshot in it.children) {
+                val answer = getAnswer(courseSnapshot)
+                if (answer != null) {
+                    answers.add(answer)
+                }
+            }
+            future.complete(answers)
+        }.addOnFailureListener {
+            future.completeExceptionally(it)
+        }
+
+        return future.get(5,TimeUnit.SECONDS)
+    }
+
+    //Note that using user.question in the main is false because you don't take new values in the db into account !
+    override fun getUserQuestions(user: User): Set<Question> {
+        val future = CompletableFuture<Set<Question>>()
+        // go in "user/userId" dir
+        db.child(usersPath).child(user.userId).child(questionsPath).get().addOnSuccessListener {
+            val questions = mutableSetOf<Question>()
+            // add every user's question that is not null in the map
+            for (courseSnapshot in it.children) {
+                val question = getQuestion(courseSnapshot)
+                if (question != null) {
+                    questions.add(question)
+                }
+            }
+            future.complete(questions)
+        }.addOnFailureListener {
+            future.completeExceptionally(it)
+        }
+
+        return future.get(5,TimeUnit.SECONDS)
+    }
+
+    override fun getUserAnswers(user: User): Set<Answer> {
+        val future = CompletableFuture<Set<Answer>>()
+        // go in "user/userId" dir
+        db.child(usersPath).child(user.userId).child(answersPath).get().addOnSuccessListener {
+            val answers = mutableSetOf<Answer>()
+            // add every user's question that is not null in the map
+            for (courseSnapshot in it.children) {
+                val answer = getAnswer(courseSnapshot)
+                if (answer != null) {
+                    answers.add(answer)
+                }
+            }
+            future.complete(answers)
+        }.addOnFailureListener {
+            future.completeExceptionally(it)
+        }
+
+        return future.get(5,TimeUnit.SECONDS)
```

Comment:
Maybe it would be a good thing to declare a constante for the "5" because it is used in several functions